lib/screens/screens_by_alert/gen_server.ex

```diff
@@ -40,22 +48,133 @@ defmodule Screens.ScreensByAlert.GenServer do
+          screens_last_updated_ttl_seconds: screens_last_updated_ttl_seconds
+        } = state
+      ) do
+    updated_screens_by_alert =
+      alert_ids
+      |> Enum.map(fn alert_id ->
+        # Check if object should expire in #{screens_by_alert_ttl_seconds} seconds
+        Process.send_after(self(), {:expire_alert, alert_id}, screens_by_alert_ttl_seconds * 1000)
+        {alert_id, %{screen_ids: [screen_id], created_at: System.system_time(:second)}}
+      end)
+      |> Map.new()
+      # Combine list of screen_ids if alert already exists.
+      # Otherwise, create a new alert.
+      |> Map.merge(screens_by_alert, fn
+        _, %{screen_ids: screens1, created_at: created_at}, %{screen_ids: screens2} ->
+          %{screen_ids: Enum.uniq(screens1 ++ screens2), created_at: created_at}
+      end)
+
+    # Overwrites previous last_updated if screen_id is already in map.
+    updated_screens_last_updated =
+      Map.put(screens_last_updated, screen_id, System.system_time(:second))
+
+    # Check if object should expire in #{screens_last_updated_ttl_seconds} seconds
+    Process.send_after(
+      self(),
+      {:expire_last_updated, screen_id},
+      screens_last_updated_ttl_seconds * 1000
+    )
+
+    {:reply, :ok,
+     %__MODULE__{
+       state
+       | screens_by_alert: updated_screens_by_alert,
+         screens_last_updated: updated_screens_last_updated
+     }}
   end
 
   @impl GenServer
-  def handle_call({:get_screens_by_alert, _alert_id}, _from, state) do
-    {:reply, [], state}
+  def handle_call(
+        {:get_screens_by_alert, alert_id},
+        _from,
+        %__MODULE__{
+          screens_by_alert: screens_by_alert,
+          screens_last_updated: screens_last_updated
+        } = state
+      ) do
+    result =
+      case Map.get(screens_by_alert, alert_id) do
+        %{screen_ids: screen_ids} ->
+          Enum.map(screen_ids, fn
+            screen_id when screen_ids != [] ->
+              last_updated = Map.get(screens_last_updated, screen_id)
+              {screen_id, last_updated}
+          end)
+
+        # Alert either expired or never existed.
+        _ ->
+          []
+      end
+
+    {:reply, result, state}
   end
 
   @impl GenServer
-  def handle_call({:get_screens_last_updated, _screen_id}, _from, state) do
-    {:reply, 0, state}
+  def handle_call(
+        {:get_screens_last_updated, screen_id},
+        _from,
+        %__MODULE__{
+          screens_last_updated: screens_last_updated
+        } = state
+      ) do
+    {:reply, Map.get(screens_last_updated, screen_id), state}
+  end
+
+  @impl GenServer
+  def handle_info(
+        {:expire_alert, alert_id},
+        %__MODULE__{screens_by_alert: screens_by_alert, screens_by_alert_ttl_seconds: ttl_seconds} =
+          state
+      ) do
+    new_state =
+      case Map.get(screens_by_alert, alert_id) do
+        %{created_at: created_at} ->
+          if created_at + ttl_seconds <= System.system_time(:second) do
```

Comment:
Ohhh, I see now—so the expiration doesn't happen if `created_at` was updated between the timer being set and the timer triggering. That works. If you stick with this approach, consider renaming the tag for this message to `:maybe_expire_alert` to make it clearer that it doesn't always result in an expiration.